feat(docs-infra): add close button to search-results aio panel

dario-piotrowicz opened the pull request:
add a close button to the search-results aio panel so that the user can
conveniently close it via keyboard

this complements the focus trap implemented in PR #44989
(more here: https://github.com/angular/angular/pull/44989#issuecomment-1037287678)

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [x] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## Issue

Issue Number: N/A

## Change
![Peek 2022-04-09 20-42](https://user-images.githubusercontent.com/61631103/162589370-e585c91b-d87c-4a27-8fd1-819ae09fdcba.gif)


![Peek 2022-04-09 20-43](https://user-images.githubusercontent.com/61631103/162589366-5b0fa183-eafc-467f-b148-a821dda85ec9.gif)



## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information
 - @gkalpak :smiley:

dario-piotrowicz committed:
feat(docs-infra): add close button to search-results aio panel

add a close button to the search-results aio panel so that the user can
conveniently close it via keyboard

this complements the focus trap implemented in PR #44989
(more here: https://github.com/angular/angular/pull/44989#issuecomment-1037287678)

dario-piotrowicz commented:
@gkalpak let me know what you think :slightly_smiling_face:

I went for a sort of minimalistic implementation, you only see the button when you may actually want to use it, otherwise it's no there at all

I started with the button always being there, and it worked pretty ok for the most part, but on some window sizes the button could overlap some of the text in the panel, which just looked ugly, so I thought that this current implementation would be nicer instead (and if the button overlaps some text in some cases it's ok here)

dario-piotrowicz commented on aio/src/app/app.component.ts:
```diff
@@ -295,13 +295,10 @@ export class AppComponent implements OnInit {
-            anchors[anchors.length - 1].focus();
-          }
+          // loop it back to the results' close button
+          const closeBtn: HTMLButtonElement =
+            this.searchResultsView.nativeElement.querySelector('button.close-btn');
```
using `querySelector` to get the button if pretty ugly I think, but it is more or less as ugly as the old code, so hopefully it's ok :sweat:

dario-piotrowicz commented on aio/src/app/shared/search-results/search-results.component.html:
```diff
@@ -52,6 +52,12 @@ <h3 class="search-section-header">{{area.name}} ({{area.pages.length + area.prio
     </div>
   </ng-container>
+
+  <div class="close-btn-wrapper">
```
the wrapper is needed for moving the button off screen when it is not in focus (with an overflow of hidden)

(I could not do that on the main search results div as that needs to be scrollable)

gkalpak replied:
Couldn't we use `display: none` on the button or something?

dario-piotrowicz replied:
I actually just checked and noticed that the wrapper is not even necessary anymore, it was at the beginning whilst I was experimenting with stuff but now everything actually works nicely without it

sorry! :bow: :cry: (and thanks a lot for pointing it out! :heart:)

dario-piotrowicz commented on aio/src/styles/2-modules/notification/_notification.scss:
```diff
@@ -83,6 +83,10 @@ aio-notification {
   .search-results {
     padding-top: $notificationHeight;
+
+    .close-btn-wrapper {
+      top: 6.5rem;
+    }
```
this is needed otherwise the button would not be visible when there is a notification :sweat: 

![Screenshot at 2022-04-09 21-24-56](https://user-images.githubusercontent.com/61631103/162590583-327beec9-328b-44e3-8b24-bc7bb5765b86.png)

gkalpak commented on aio/src/app/shared/shared.module.ts:
```diff
@@ -2,10 +2,12 @@ import { NgModule } from '@angular/core';
   imports: [
-    CommonModule
+    CommonModule,
+    MatIconModule,
```
If we're adding `MatIconModule` here, should we take it off of the [AppModule][1]?

[1]: https://github.com/angular/angular/blob/e0ac61412137144a43dc5a2134b4cd620bb4c30f/aio/src/app/app.module.ts#L153

dario-piotrowicz replied:
to remove it from the `AppModule` I would need to export it here in the `SharedModule`, I am not completely sure if I like it

it seems like it makes a very small difference in the main bundle size, I can do it if you think it's beneficial :slightly_smiling_face:

gkalpak replied:
I don't see why _not_ to do it. Why don't you like it?

dario-piotrowicz replied:
I am not sure, it feels like we are extending the module with extra stuff that it should maybe not care about :thinking: 

but yeah I am not a modules expert if you say it's fine I am too :smile:

dario-piotrowicz replied:
done :slightly_smiling_face:

mary-poppins commented:
You can preview cad8b44 at https://pr45579-cad8b44.ngbuilds.io/.
You can preview 1a062fc at https://pr45579-1a062fc.ngbuilds.io/.
You can preview b08100a at https://pr45579-b08100a.ngbuilds.io/.
You can preview 7085424 at https://pr45579-7085424.ngbuilds.io/.

gkalpak commented on aio/src/styles/2-modules/notification/_notification.scss:
```diff
@@ -83,6 +83,10 @@ aio-notification {
 
   .search-results {
     padding-top: $notificationHeight;
```
Could we change this from `padding-top` to `margin-top` and get rid of the `.close-btn-wrapper` override below?

gkalpak replied:
Ah, good point :smiley:

gkalpak reviewed (changes requested):
Thx for working on this, @dario-piotrowicz :+1:
I left a couple of comments, but overall this looks to be heading the right direction.

dario-piotrowicz commented on aio/src/styles/2-modules/search-results/_search-results.scss:
```diff
@@ -1,4 +1,5 @@
 @use '../../mixins';
+@use '../../constants';
```
sorry, classic mistake of mine :sweat:

dario-piotrowicz replied:
done :slightly_smiling_face:

dario-piotrowicz commented on aio/src/app/shared/search-results/search-results.component.html:
```diff
@@ -52,6 +52,12 @@ <h3 class="search-section-header">{{area.name}} ({{area.pages.length + area.prio
   </ng-container>
+
+  <div class="close-btn-wrapper">
+    <button class="close-btn" aria-label="Close search results panel" (click)="onCloseClicked()">
+      <mat-icon>close</mat-icon>
+    </button>
```
sure, good point :slightly_smiling_face:

dario-piotrowicz replied:
done :slightly_smiling_face:

dario-piotrowicz committed:
fixup! feat(docs-infra): add close button to search-results aio panel

align the button with the notification one

dario-piotrowicz committed:
fix(docs-infra): amend notification close button aria-label

move the aria-label used inside the close-button to the button itself
(since otherwise the button's aria-label overrides the childs)

dario-piotrowicz commented on aio/src/app/layout/notification/notification.component.html:
```diff
@@ -2,6 +2,6 @@
 
-<button mat-icon-button class="close-button" aria-label="Close" (click)="dismiss()">
-  <mat-icon svgIcon="close" aria-label="Dismiss notification"></mat-icon>
+<button mat-icon-button class="close-button" aria-label="Dismiss notification" (click)="dismiss()">
```
@gkalpak since you linked this to me I had a look and I think that the `aria-label`s here aren't used correctly, I think that the parent's `aria-label` just overrides the childs so I figured I could fix this too :slightly_smiling_face: (unless I am mistaken/missing something?)

___

Before:
![before](https://user-images.githubusercontent.com/61631103/162812187-424ef945-e66e-47d2-85e9-a813abacc795.png)

After:
![after](https://user-images.githubusercontent.com/61631103/162812207-2f03d800-9e92-49a4-8601-97c546610e61.png)

gkalpak replied:
:100:

mary-poppins commented:
You can preview c87be9d at https://pr45579-c87be9d.ngbuilds.io/.

dario-piotrowicz committed:
fixup! feat(docs-infra): add close button to search-results aio panel

reorder scss imports

gkalpak commented on aio/src/app/shared/search-results/search-results.component.html:
```diff
@@ -52,6 +52,12 @@ <h3 class="search-section-header">{{area.name}} ({{area.pages.length + area.prio
+
+  <div class="close-button-wrapper">
+    <button mat-icon-button class="close-button" aria-label="Close search results panel" (click)="onCloseClicked()">
+      <mat-icon svgIcon="close" aria-label="Dismiss notification"></mat-icon>
```
```suggestion
      <mat-icon svgIcon="close"></mat-icon>
```

:grin:

dario-piotrowicz replied:
:facepalm:

gkalpak commented on aio/src/styles/2-modules/notification/_notification.scss:
```diff
@@ -83,6 +83,10 @@ aio-notification {
     padding-top: $notificationHeight;
+
+    .close-button-wrapper {
+      top: 6.5rem;
```
Can you express this as a function of `$notificationHeight`?
For example:

```suggestion
      top: 1rem + $notificationHeight;
```

dario-piotrowicz replied:
you made me look at `$notificationHeight` :see_no_evil: 

it uses pixels, which do no account for different font-sizes:
![Peek 2022-04-11 23-21](https://user-images.githubusercontent.com/61631103/162843105-89d1989c-410d-4b63-8800-c9ac9f2f7099.gif)

I think I'll want to fix that next :sweat_smile:

dario-piotrowicz replied:
actually `$notificationHeight` works just find :sweat_smile::+1:

dario-piotrowicz committed:
fixup! feat(docs-infra): add close button to search-results aio panel

dario-piotrowicz committed:
fixup! feat(docs-infra): add close button to search-results aio panel

remove close-button-wrapper

mary-poppins commented:
You can preview af0d745 at https://pr45579-af0d745.ngbuilds.io/.

dario-piotrowicz commented:
@gkalpak this is unrelated to the changes, I just wanted to tell you that I noticed this behavior :sweat: 

![Screenshot at 2022-04-11 23-43-31](https://user-images.githubusercontent.com/61631103/162845416-6d402158-3a7a-4ff1-8b6a-efaffef8ae8d.png)

seems lightly tricky to fix with the current implementation but I'll have a look at it (if you want :smiley:) in a follow up PR :slightly_smiling_face:

gkalpak commented:
Oh, interesting. This is only a problem on the home page (due to the transparent toolbar) and only when there is a notification visible.

I believe we can fix this replacing the extra [top padding][1] with some extra top border (on top of the regular [top border][2]. So, something like:

```diff
 // src/styles/2-modules/notification/_notification.scss
 .aio-notification-show {
   // ...
 
   .search-results {
-    padding-top: $notificationHeight;
+    border-top-width: 68px + $notificationHeight;
   }
```

This would also remove the need to adjust the close button's `top` when a notification is visible.
WDYT? (Did I miss something again? 😟)

[1]: https://github.com/angular/angular/blob/d11d1c0b70bda8a275a16ebd9a192beafcbc2a14/aio/src/styles/2-modules/notification/_notification.scss#L85
[2]: https://github.com/angular/angular/blob/d11d1c0b70bda8a275a16ebd9a192beafcbc2a14/aio/src/styles/2-modules/search-results/_search-results.scss#L11

dario-piotrowicz committed:
fixup! feat(docs-infra): add close button to search-results aio panel

remove no longer needed top for close button

dario-piotrowicz committed:
fix(docs-infra): fix aio search-results issue with toolbar and notification

use border-top-width instead of padding-top as the strategy for moving
the search-results panel down, this fixes the issue of the panel going
behind the toolbar (which causes either overlapping text in the home
page or the results scrollbar to be hidden behind the toolbar in other
pages)

dario-piotrowicz commented:
> Oh, interesting. This is only a problem on the home page (due to the transparent toolbar) and only when there is a notification visible.
> 
> I believe we can fix this replacing the extra [top padding](https://github.com/angular/angular/blob/d11d1c0b70bda8a275a16ebd9a192beafcbc2a14/aio/src/styles/2-modules/notification/_notification.scss#L85) with some extra top border (on top of the regular [top border](https://github.com/angular/angular/blob/d11d1c0b70bda8a275a16ebd9a192beafcbc2a14/aio/src/styles/2-modules/search-results/_search-results.scss#L11). So, something like:
> 
> ```diff
>  // src/styles/2-modules/notification/_notification.scss
>  .aio-notification-show {
>    // ...
>  
>    .search-results {
> -    padding-top: $notificationHeight;
> +    border-top-width: 68px + $notificationHeight;
>    }
> ```
> 
> This would also remove the need to adjust the close button's `top` when a notification is visible. WDYT? (Did I miss something again? worried)

Ah yeah I got caught up in that issue there and did not remembered that the toolbar is only transparent in the home :sweat_smile: , anyways in the other pages the issue is not as evident but you can see the toolbar covering the results scrollbar :sweat_smile: 

Anyways your solution does fix it nicely (and it's not much hackier than before :sweat_smile:) and as you said the button doesn't need the top anymore
Awesome, I've applied the change, thanks a lot :smile:

mary-poppins commented:
You can preview fe09858 at https://pr45579-fe09858.ngbuilds.io/.

gkalpak commented on aio/src/styles/2-modules/search-results/_search-results.scss:
```diff
@@ -71,6 +72,31 @@ aio-search-results {
+      right: $offset;
+      border: none;
+      padding: 0;
+      margin: 0;
```
Why not set `margin: 3px` here and not in the `&:focus` block?

dario-piotrowicz replied:
ah yeah! sorry it made sense when there was the wrapper but now it is useless! thanks a lot for noticing it! :heart::+1:

gkalpak commented on aio/src/styles/2-modules/search-results/_search-results.scss:
```diff
@@ -71,6 +72,31 @@ aio-search-results {
+      margin: 0;
+      display: flex;
+      opacity: 0;
+      height: 0;
+      width: 0;
```
OOC, why set these to 0?

dario-piotrowicz replied:
to avoid the button being clickable then it should not :slightly_smiling_face: 

![Peek 2022-04-13 20-44](https://user-images.githubusercontent.com/61631103/163257992-04376ae4-c9d5-427e-a783-d2e3ded1519a.gif)

As you can see from the gif I actually need to only set one of those to 0, sorry my bad, I'll fix that :slightly_smiling_face:

dario-piotrowicz replied:
fixed :slightly_smiling_face:

gkalpak commented on aio/src/styles/2-modules/search-results/_search-results.scss:
```diff
@@ -71,6 +72,31 @@ aio-search-results {
+      opacity: 0;
+      height: 0;
+      width: 0;
+      color: constants.$white;
+      background-color: constants.$darkgray;
```
I think it would be better to set these to the more future-proof:

```suggestion
      color: inherit;
      background: none;
```

:wink:

dario-piotrowicz replied:
I did consider those! :slightly_smiling_face: 

but the `color: inherit` does not work because the white of the text does not come from the panel but on the elements themselves, the panel doesn't have a color I think so if we try to inherit it we get nothing (and I didn't want to mess around with things)

about the background I added it because in some specific cases the button can be above the text so I wanted to prevent any possible visual bug :slightly_smiling_face: , here I can actually use `inherit` that slipped my mind :sweat: :bow: 

![Peek 2022-04-13 21-01](https://user-images.githubusercontent.com/61631103/163261368-38df4787-0a78-4047-84a9-e182585c6f37.gif)

dario-piotrowicz replied:
actually the white comes from two different places (with two slightly different whites):

![Screenshot at 2022-04-13 21-12-06](https://user-images.githubusercontent.com/61631103/163262761-a5c1acdd-02e1-4cb0-9247-a6cbc85cc389.png)

![Screenshot at 2022-04-13 21-11-56](https://user-images.githubusercontent.com/61631103/163262780-381b4d0b-4dff-46b5-a40b-2de3d09d4792.png)

I can refactor some of those places to include the button as well :thinking:

gkalpak replied:
I see. Yeah, let's use `inherit` for the `background-color` and update the `color` styles to include the close button (so we only have to update the `color` in one place if we decide to change it in the future).

dario-piotrowicz replied:
done, I've updated the colors :slightly_smiling_face: 

for the color itself I needed to introduce a sass variable in order not to mess up the nesting there, let me know if that's ok :slightly_smiling_face:

gkalpak reviewed (changes requested):
A couple of minor styling suggestions. Otherwise lgtm 🚀

dario-piotrowicz committed:
fixup! feat(docs-infra): add close button to search-results aio panel

mary-poppins commented:
You can preview 36db6ca at https://pr45579-36db6ca.ngbuilds.io/.

mary-poppins commented:
You can preview 8b51b1d at https://pr45579-8b51b1d.ngbuilds.io/.

dario-piotrowicz committed:
fixup! feat(docs-infra): add close button to search-results aio panel

use item-color for close-button

mary-poppins commented:
You can preview 75f56b0 at https://pr45579-75f56b0.ngbuilds.io/.

gkalpak commented on aio/src/styles/2-modules/search-results/_search-results.scss:
```diff
@@ -1,3 +1,4 @@
+@use '../../constants';
```
I believe this import is no longer needed.

dario-piotrowicz replied:
:man_facepalming: sorry I'm on it! :cry:

dario-piotrowicz replied:
removed, sorry again :sweat: :cry:

gkalpak reviewed (approved):
One last comment. Otherwise lgtm 🚀
Thx for working through this, @dario-piotrowicz ❤️

AndrewKushnir reviewed (approved):
Reviewed-for: size-tracking

atscott reviewed (approved):
reviewed-for: size-tracking

dylhunn commented:
@dario-piotrowicz Ping me when this is ready to merge.

dario-piotrowicz committed:
fixup! feat(docs-infra): add close button to search-results aio panel

remove no longer needed use statement

mary-poppins commented:
You can preview fc26acc at https://pr45579-fc26acc.ngbuilds.io/.

dylhunn commented:
This PR was merged into the repository by commit 0b6407241c68536c1424201e767628d507760e71.
